File: frontend/src/api/feeds.js
/**
 * ===================================================================
 * 소셜 피드 관리 API
 * ===================================================================
 *
 * 주요 기능:
 * - 피드 작성/수정/삭제 (이미지 업로드 지원)
 * - 피드 목록 조회 및 페이지네이션
 * - 피드 좋아요 토글
 * - 피드 댓글 시스템 (댓글/대댓글)
 * - 내가 작성한 피드 조회
 * - 피드 검색 기능
 *
 * 기술 스택:
 * - HTTP Client: Axios
 * - 파일 업로드: FormData (multipart/form-data)
 * - 인증: 세션 쿠키 기반
 *
 * 사용 예시:
 * - 메인 피드 페이지에서 게시물 목록 표시
 * - 가상 피팅 결과를 피드로 공유
 * - 의류 코디네이션 팁 공유
 * - 사용자 간 소셜 인터랙션 (좋아요, 댓글)
 * - 마이페이지에서 내 게시물 관리
 *
 * 이미지 처리:
 * - 다중 이미지 업로드 지원
 * - 이미지 순서 관리
 * - FormData를 통한 파일 전송
 * ===================================================================
 */

import axios from "axios"

// API 기본 설정
const API_BASE_URL = process.env.REACT_APP_API_URL || "http://localhost:8000"

// axios 인스턴스 생성
const feedAPI = axios.create({
  baseURL: `${API_BASE_URL}/api/feeds`,
  withCredentials: true, // 세션 쿠키 전송
  timeout: 30000, // 파일 업로드를 위해 타임아웃 증가
})

// 요청 인터셉터
feedAPI.interceptors.request.use(
  (config) => {
    console.log(`📤 API 요청: ${config.method?.toUpperCase()} ${config.url}`, config.data)
    return config
  },
  (error) => {
    console.error("📤 요청 오류:", error)
    return Promise.reject(error)
  },
)

// 응답 인터셉터
feedAPI.interceptors.response.use(
  (response) => {
    console.log(`📥 API 응답: ${response.config.method?.toUpperCase()} ${response.config.url}`, response.data)
    return response
  },
  (error) => {
    console.error("📥 응답 오류:", error)

    // 상세한 오류 정보 로깅
    if (error.response) {
      console.error("응답 상태:", error.response.status)
      console.error("응답 데이터:", error.response.data)
      console.error("응답 헤더:", error.response.headers)
    } else if (error.request) {
      console.error("요청 오류:", error.request)
    } else {
      console.error("설정 오류:", error.message)
    }

    // 인증 오류 처리
    if (error.response?.status === 401) {
      console.error("인증이 필요합니다. 로그인 페이지로 이동해주세요.")
    }

    return Promise.reject(error)
  },
)

/**
 * 피드 작성
 * @param {Object} feedData - 피드 데이터 (title, content)
 * @param {Array} imageFiles - 이미지 파일 배열
 * @returns {Promise<Object>} - 생성된 피드 정보
 */
export const createFeed = async (feedData, imageFiles = []) => {
  try {
    // FormData 생성
    const formData = new FormData()
    formData.append("title", feedData.title)
    formData.append("content", feedData.content)

    // 이미지 파일과 순서 정보 추가
    imageFiles.forEach((image, index) => {
      if (image.file) {
        formData.append("images", image.file)
        formData.append("image_orders", image.order || index + 1)
      }
    })

    const response = await feedAPI.post("/", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    })

    return response.data
  } catch (error) {
    console.error("피드 작성 실패:", error)
    throw error
  }
}

/**
 * 피드 목록 조회
 * @param {Object} params - 조회 파라미터
 * @returns {Promise<Object>} - 피드 목록과 페이지네이션 정보
 */
export const getFeeds = async (params = {}) => {
  try {
    const { page = 1, size = 10, sort_by = "created_at", order = "desc" } = params

    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sort_by,
      order,
    })

    const response = await feedAPI.get(`/?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("피드 목록 조회 실패:", error)
    throw error
  }
}

/**
 * 특정 피드 조회
 * @param {number} feedId - 피드 ID
 * @returns {Promise<Object>} - 피드 정보
 */
export const getFeedById = async (feedId) => {
  try {
    const response = await feedAPI.get(`/${feedId}`)
    return response.data
  } catch (error) {
    console.error("Feed lookup failed:", error)
    throw error
  }
}

/**
 * 피드 수정
 * @param {number} feedId - 피드 ID
 * @param {Object} feedData - 수정할 피드 데이터
 * @param {Array} imageFiles - 새로운 이미지 파일 배열
 * @returns {Promise<Object>} - 수정된 피드 정보
 */
export const updateFeed = async (feedId, feedData, imageFiles = [], existingImageIds = '') => {
  try {
    const formData = new FormData()
    formData.append("title", feedData.title)
    formData.append("content", feedData.content)

    // 유지할 기존 이미지 ID들 추가
    if (existingImageIds) {
      formData.append("existing_image_ids", existingImageIds)
    }

    // 이미지 파일과 순서 정보 추가
    imageFiles.forEach((image, index) => {
      if (image.file) {
        formData.append("images", image.file)
        formData.append("image_orders", image.order || index + 1)
      }
    })

    const response = await feedAPI.put(`/${feedId}`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    })

    return response.data
  } catch (error) {
    console.error("피드 수정 실패:", error)
    throw error
  }
}

/**
 * 피드 삭제
 * @param {number} feedId - 피드 ID
 * @returns {Promise<Object>} - 삭제 결과
 */
export const deleteFeed = async (feedId) => {
  try {
    const response = await feedAPI.delete(`/${feedId}`)
    return response.data
  } catch (error) {
    console.error("피드 삭제 실패:", error)
    throw error
  }
}

/**
 * 피드 좋아요 토글
 * @param {number} feedId - 피드 ID
 * @returns {Promise<Object>} - 좋아요 상태 정보
 */
export const toggleFeedLike = async (feedId) => {
  try {
    const response = await feedAPI.post(`/${feedId}/like`)
    return response.data
  } catch (error) {
    console.error("피드 좋아요 토글 실패:", error)
    throw error
  }
}

/**
 * 피드 댓글 목록 조회
 * @param {number} feedId - 피드 ID
 * @param {Object} params - 조회 파라미터
 * @returns {Promise<Object>} - 댓글 목록
 */
export const getFeedComments = async (feedId, params = {}) => {
  try {
    const { page = 1, size = 20, tree_structure = true } = params
    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      tree_structure: tree_structure.toString(),
    })

    // 경로 끝에 슬래시 추가
    const response = await feedAPI.get(`/${feedId}/comments/?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("피드 댓글 조회 실패:", error)
    throw error
  }
}

/**
 * 피드 댓글 작성
 * @param {number} feedId - 피드 ID
 * @param {string} content - 댓글 내용
 * @param {number|null} parentId - 부모 댓글 ID (대댓글인 경우)
 * @returns {Promise<Object>} - 생성된 댓글 정보
 */
export const createFeedComment = async (feedId, content, parentId = null) => {
  try {
    console.log(`💬 댓글 작성 API 호출: feedId=${feedId}, content="${content}", parentId=${parentId}`)

    const requestData = {
      content: content.trim(),
    }

    // parentId가 유효한 값인 경우에만 추가
    if (parentId && parentId > 0) {
      requestData.parent_id = parentId
    }

    console.log(`📤 요청 데이터:`, requestData)

    // 경로 끝에 슬래시 추가
    const response = await feedAPI.post(`/${feedId}/comments/`, requestData, {
      headers: {
        "Content-Type": "application/json",
      },
    })

    console.log(`✅ 댓글 작성 성공:`, response.data)
    return response.data
  } catch (error) {
    console.error("피드 댓글 작성 실패:", error)

    // 더 상세한 오류 정보 로깅
    if (error.response?.data) {
      console.error("서버 오류 응답:", error.response.data)
    }

    throw error
  }
}

/**
 * 피드 댓글 수정
 * @param {number} feedId - 피드 ID
 * @param {number} commentId - 댓글 ID
 * @param {string} content - 수정할 댓글 내용
 * @returns {Promise<Object>} - 수정된 댓글 정보
 */
export const updateFeedComment = async (feedId, commentId, content) => {
  try {
    // 경로 끝에 슬래시 추가
    const response = await feedAPI.put(
      `/${feedId}/comments/${commentId}/`,
      {
        content: content.trim(),
      },
      {
        headers: {
          "Content-Type": "application/json",
        },
      },
    )
    return response.data
  } catch (error) {
    console.error("피드 댓글 수정 실패:", error)
    throw error
  }
}

/**
 * 피드 댓글 삭제
 * @param {number} feedId - 피드 ID
 * @param {number} commentId - 댓글 ID
 * @returns {Promise<Object>} - 삭제 결과
 */
export const deleteFeedComment = async (feedId, commentId) => {
  try {
    // 경로 끝에 슬래시 추가
    const response = await feedAPI.delete(`/${feedId}/comments/${commentId}/`)
    return response.data
  } catch (error) {
    console.error("피드 댓글 삭제 실패:", error)
    throw error
  }
}

/**
 * 특정 댓글 조회
 * @param {number} feedId - 피드 ID
 * @param {number} commentId - 댓글 ID
 * @returns {Promise<Object>} - 댓글 정보
 */
export const getFeedComment = async (feedId, commentId) => {
  try {
    // 경로 끝에 슬래시 추가
    const response = await feedAPI.get(`/${feedId}/comments/${commentId}/`)
    return response.data
  } catch (error) {
    console.error("피드 댓글 조회 실패:", error)
    throw error
  }
}

/**
 * 댓글의 대댓글 목록 조회
 * @param {number} feedId - 피드 ID
 * @param {number} commentId - 부모 댓글 ID
 * @param {Object} params - 조회 파라미터
 * @returns {Promise<Object>} - 대댓글 목록
 */
export const getCommentReplies = async (feedId, commentId, params = {}) => {
  try {
    const { page = 1, size = 50 } = params
    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    })

    // 경로 끝에 슬래시 추가
    const response = await feedAPI.get(`/${feedId}/comments/${commentId}/replies/?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("대댓글 목록 조회 실패:", error)
    throw error
  }
}

/**
 * 내가 작성한 피드 목록 조회
 * @param {Object} params - 조회 파라미터
 * @returns {Promise<Object>} - 내 피드 목록
 */
export const getMyFeeds = async (params = {}) => {
  try {
    const { page = 1, size = 10 } = params
    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    })

    console.log("📤 getMyFeeds 요청:", {
      url: `/my-feeds?${queryParams}`,
      params,
    })

    const response = await feedAPI.get(`/my-feeds?${queryParams}`)

    console.log("📥 getMyFeeds 응답:", response.data)
    return response.data
  } catch (error) {
    console.error("❌ 내 피드 목록 조회 실패:", error)

    // 더 상세한 에러 정보
    if (error.response) {
      console.error("에러 상태:", error.response.status)
      console.error("에러 데이터:", error.response.data)
      console.error("에러 헤더:", error.response.headers)
    }

    throw error
  }
}

/**
 * 피드 검색
 * @param {string} query - 검색어
 * @param {Object} params - 검색 파라미터
 * @returns {Promise<Object>} - 검색 결과
 */
export const searchFeeds = async (query, params = {}) => {
  try {
    const { page = 1, size = 10 } = params
    const queryParams = new URLSearchParams({
      q: query,
      page: page.toString(),
      size: size.toString(),
    })

    const response = await feedAPI.get(`/search?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("피드 검색 실패:", error)
    throw error
  }
}

File: backend/app/api/routes/feeds.py
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File, Form, Request
from sqlalchemy.orm import Session
from typing import List, Optional
import os
import uuid
from datetime import datetime

from app.db.database import get_db
from app.models.users import Users
from app.models.feeds import Feeds
from app.models.feed_images import FeedImages
from app.schemas.feeds import FeedCreate, FeedResponse, FeedListResponse
from app.api.dependencies import get_current_user
from app.utils.file_upload import save_upload_file

router = APIRouter(
    prefix="/api/feeds",
    tags=["feeds"]
)

def get_feed_with_user_status(feed, current_user_id, db):
    """피드 정보에 현재 사용자의 상태 정보를 추가하는 헬퍼 함수"""
    user = db.query(Users).filter(Users.user_id == feed.user_id).first()
    images = db.query(FeedImages).filter(FeedImages.feed_id == feed.feed_id).order_by(FeedImages.image_order).all()
    
    # 기본값 설정
    like_count = 0
    comment_count = 0
    is_liked = False
    is_following = False
    
    # 좋아요 수와 현재 사용자의 좋아요 상태 확인
    try:
        from app.models.liked_feeds import LikedFeeds
        
        # 전체 좋아요 수 계산
        like_count = db.query(LikedFeeds).filter(LikedFeeds.feed_id == feed.feed_id).count()
        
        # 현재 사용자의 좋아요 상태 확인 (로그인한 경우에만)
        if current_user_id:
            user_like = db.query(LikedFeeds).filter(
                LikedFeeds.feed_id == feed.feed_id,
                LikedFeeds.user_id == current_user_id
            ).first()
            is_liked = user_like is not None
            
    except ImportError:
        pass
    except Exception:
        pass
    
    # 댓글 수 계산
    try:
        from app.models.feed_comments import FeedComments
        comment_count = db.query(FeedComments).filter(FeedComments.feed_id == feed.feed_id).count()
    except ImportError:
        pass
    except Exception:
        pass
    
    # 팔로우 상태 확인
    if current_user_id and user and current_user_id != user.user_id:
        try:
            from app.models.users import Followers
            follow_record = db.query(Followers).filter(
                Followers.follower_id == current_user_id,
                Followers.following_id == user.user_id
            ).first()
            is_following = follow_record is not None
        except ImportError:
            # Followers 모델이 없는 경우
            try:
                from app.crud.followers import is_following as check_following
                is_following = check_following(db, current_user_id, user.user_id)
            except ImportError:
                pass
        except Exception:
            pass
    
    result = {
        "feed_id": feed.feed_id,
        "user_id": feed.user_id,
        "title": feed.title,
        "content": feed.content,
        "created_at": feed.created_at,
        "updated_at": feed.updated_at,
        "user": {
            "user_id": user.user_id,
            "nickname": user.nickname,
            "email": user.email,  # 이메일 정보 추가
            "profile_picture": user.profile_picture,
            "isFollowing": is_following  # 팔로우 상태 추가
        } if user else None,
        "images": [
            {
                "id": img.id,
                "image_url": img.image_url,
                "image_order": img.image_order
            }
            for img in images
        ],
        "like_count": like_count,
        "comment_count": comment_count,
        "is_liked": is_liked
    }
    
    return result

# 피드 작성
@router.post("/", response_model=FeedResponse)
async def create_feed(
    title: str = Form(...),
    content: str = Form(...),
    images: List[UploadFile] = File(None),
    image_orders: List[int] = Form(None),
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """피드 작성 API"""
    try:
        # 피드 생성
        new_feed = Feeds(
            user_id=current_user.user_id,
            title=title,
            content=content,
            created_at=datetime.now(),
            updated_at=datetime.now()
        )
        db.add(new_feed)
        db.flush()
        
        # 이미지 처리
        if images and image_orders:
            for i, image_file in enumerate(images):
                file_ext = os.path.splitext(image_file.filename)[1]
                unique_filename = f"{uuid.uuid4()}{file_ext}"
                upload_dir = os.path.join("uploads", "feeds")
                os.makedirs(upload_dir, exist_ok=True)
                file_path = os.path.join(upload_dir, unique_filename)
                
                await save_upload_file(image_file, file_path)
                
                image_url = f"/uploads/feeds/{unique_filename}"
                
                feed_image = FeedImages(
                    feed_id=new_feed.feed_id,
                    image_url=image_url,
                    image_order=image_orders[i] if i < len(image_orders) else i + 1,
                    created_at=datetime.now()
                )
                db.add(feed_image)
        
        db.commit()
        db.refresh(new_feed)
        
        response_data = get_feed_with_user_status(new_feed, current_user.user_id, db)
        return response_data
    
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 작성 중 오류가 발생했습니다: {str(e)}"
        )

# 피드 목록 조회
@router.get("/", response_model=FeedListResponse)
async def get_feeds(
    page: int = 1,
    size: int = 10,
    sort_by: str = "created_at",
    order: str = "desc",
    db: Session = Depends(get_db),
    request: Request = None
):
    """피드 목록 조회 API"""
    try:
        # 현재 사용자 확인
        current_user_id = None
        try:
            if request and hasattr(request, 'session') and 'user_id' in request.session:
                user_id = request.session['user_id']
                current_user = db.query(Users).filter(Users.user_id == user_id).first()
                if current_user:
                    current_user_id = current_user.user_id
        except Exception:
            pass
        
        # 페이지네이션 설정
        skip = (page - 1) * size
        
        # 정렬 설정
        order_by = getattr(Feeds, sort_by, Feeds.created_at)
        if order.lower() == "desc":
            order_by = order_by.desc()
        else:
            order_by = order_by.asc()
        
        # 피드 목록 조회
        feeds = db.query(Feeds).order_by(order_by).offset(skip).limit(size).all()
        total = db.query(Feeds).count()
        
        # 응답 데이터 준비
        feed_list = []
        for feed in feeds:
            feed_data = get_feed_with_user_status(feed, current_user_id, db)
            feed_list.append(feed_data)
        
        return {
            "feeds": feed_list,
            "total": total,
            "page": page,
            "size": size,
            "total_pages": (total + size - 1) // size
        }
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 목록 조회 중 오류가 발생했습니다: {str(e)}"
        )
        
# 내가 작성한 피드 목록 조회
@router.get("/my-feeds", response_model=FeedListResponse)
async def get_my_feeds(
    page: int = 1,
    size: int = 10,
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """내가 작성한 피드 목록 조회 API"""
    try:
        skip = (page - 1) * size
        
        feeds = db.query(Feeds).filter(
            Feeds.user_id == current_user.user_id
        ).order_by(Feeds.created_at.desc()).offset(skip).limit(size).all()
        
        total = db.query(Feeds).filter(Feeds.user_id == current_user.user_id).count()
        
        feed_list = []
        for feed in feeds:
            feed_data = get_feed_with_user_status(feed, current_user.user_id, db)
            feed_list.append(feed_data)
        
        return {
            "feeds": feed_list,
            "total": total,
            "page": page,
            "size": size,
            "total_pages": (total + size - 1) // size
        }
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"내 피드 목록 조회 중 오류가 발생했습니다: {str(e)}"
        )
        
# 피드 검색
@router.get("/search", response_model=FeedListResponse)
async def search_feeds(
    q: str,
    page: int = 1,
    size: int = 10,
    db: Session = Depends(get_db),
    request: Request = None
):
    """피드 검색 API"""
    try:
        skip = (page - 1) * size
        
        # 제목이나 내용에서 검색
        feeds = db.query(Feeds).filter(
            (Feeds.title.contains(q)) | (Feeds.content.contains(q))
        ).order_by(Feeds.created_at.desc()).offset(skip).limit(size).all()
        
        total = db.query(Feeds).filter(
            (Feeds.title.contains(q)) | (Feeds.content.contains(q))
        ).count()
        
        # 현재 사용자 확인
        current_user_id = None
        try:
            if request and hasattr(request, 'session') and 'user_id' in request.session:
                current_user_id = request.session['user_id']
        except Exception:
            pass
        
        feed_list = []
        for feed in feeds:
            feed_data = get_feed_with_user_status(feed, current_user_id, db)
            feed_list.append(feed_data)
        
        return {
            "feeds": feed_list,
            "total": total,
            "page": page,
            "size": size,
            "total_pages": (total + size - 1) // size
        }
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 검색 중 오류가 발생했습니다: {str(e)}"
        )

# 특정 피드 조회
@router.get("/{feed_id}", response_model=FeedResponse)
async def get_feed(
    feed_id: int,
    db: Session = Depends(get_db),
    request: Request = None
):
    """특정 피드 조회 API"""
    try:
        feed = db.query(Feeds).filter(Feeds.feed_id == feed_id).first()
        if not feed:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="피드를 찾을 수 없습니다."
            )
        
        # 현재 사용자 확인
        current_user_id = None
        try:
            if request and hasattr(request, 'session') and 'user_id' in request.session:
                current_user_id = request.session['user_id']
        except Exception:
            pass
        
        response_data = get_feed_with_user_status(feed, current_user_id, db)
        return response_data
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 조회 중 오류가 발생했습니다: {str(e)}"
        )

# 피드 수정
@router.put("/{feed_id}", response_model=FeedResponse)
async def update_feed(
    feed_id: int,
    title: str = Form(...),
    content: str = Form(...),
    images: List[UploadFile] = File(None),
    image_orders: List[int] = Form(None),
    existing_image_ids: str = Form(None),  # 유지할 기존 이미지 ID들 (쉼표로 구분)
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """피드 수정 API"""
    try:
        # 피드 존재 확인 및 권한 확인
        feed = db.query(Feeds).filter(Feeds.feed_id == feed_id).first()
        if not feed:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="피드를 찾을 수 없습니다."
            )
        
        if feed.user_id != current_user.user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="피드를 수정할 권한이 없습니다."
            )
        
        # 피드 정보 업데이트
        feed.title = title
        feed.content = content
        feed.updated_at = datetime.now()
        
        # 이미지 처리 로직 개선
        # 1. 유지할 기존 이미지 ID 파싱
        keep_image_ids = []
        if existing_image_ids:
            try:
                keep_image_ids = [int(id.strip()) for id in existing_image_ids.split(',') if id.strip()]
            except ValueError:
                keep_image_ids = []
        
        # 2. 기존 이미지 중 삭제할 이미지들 제거
        existing_images = db.query(FeedImages).filter(FeedImages.feed_id == feed_id).all()
        for img in existing_images:
            if img.id not in keep_image_ids:
                try:
                    file_path = img.image_url.replace("/uploads/", "uploads/")
                    if os.path.exists(file_path):
                        os.remove(file_path)
                except Exception:
                    pass
                db.delete(img)
        
        # 3. 유지할 기존 이미지들의 순서 업데이트
        if keep_image_ids:
            for i, image_id in enumerate(keep_image_ids):
                existing_img = db.query(FeedImages).filter(
                    FeedImages.id == image_id,
                    FeedImages.feed_id == feed_id
                ).first()
                if existing_img:
                    existing_img.image_order = i + 1
        
        # 4. 새 이미지 저장
        if images:
            start_order = len(keep_image_ids) + 1
            for i, image_file in enumerate(images):
                file_ext = os.path.splitext(image_file.filename)[1]
                unique_filename = f"{uuid.uuid4()}{file_ext}"
                upload_dir = os.path.join("uploads", "feeds")
                os.makedirs(upload_dir, exist_ok=True)
                file_path = os.path.join(upload_dir, unique_filename)
                
                await save_upload_file(image_file, file_path)
                
                image_url = f"/uploads/feeds/{unique_filename}"
                
                feed_image = FeedImages(
                    feed_id=feed_id,
                    image_url=image_url,
                    image_order=image_orders[i] if image_orders and i < len(image_orders) else start_order + i,
                    created_at=datetime.now()
                )
                db.add(feed_image)
        
        db.commit()
        db.refresh(feed)
        
        response_data = get_feed_with_user_status(feed, current_user.user_id, db)
        return response_data
    
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 수정 중 오류가 발생했습니다: {str(e)}"
        )

# 피드 삭제
@router.delete("/{feed_id}")
async def delete_feed(
    feed_id: int,
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """피드 삭제 API"""
    try:
        # 피드 존재 확인 및 권한 확인
        feed = db.query(Feeds).filter(Feeds.feed_id == feed_id).first()
        if not feed:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="피드를 찾을 수 없습니다."
            )
        
        if feed.user_id != current_user.user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="피드를 삭제할 권한이 없습니다."
            )
        
        # 관련 이미지 파일 삭제
        images = db.query(FeedImages).filter(FeedImages.feed_id == feed_id).all()
        for img in images:
            try:
                file_path = img.image_url.replace("/uploads/", "uploads/")
                if os.path.exists(file_path):
                    os.remove(file_path)
            except Exception:
                pass
        
        # 피드 삭제
        db.delete(feed)
        db.commit()
        
        return {"message": "피드가 성공적으로 삭제되었습니다."}
    
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"피드 삭제 중 오류가 발생했습니다: {str(e)}"
        )

# 피드 좋아요 토글
@router.post("/{feed_id}/like")
async def toggle_feed_like(
    feed_id: int,
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """피드 좋아요 토글 API"""
    try:
        from app.models.liked_feeds import LikedFeeds
        
        # 피드 존재 확인
        feed = db.query(Feeds).filter(Feeds.feed_id == feed_id).first()
        if not feed:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="피드를 찾을 수 없습니다."
            )
        
        # 좋아요 여부 확인
        liked = db.query(LikedFeeds).filter(
            LikedFeeds.feed_id == feed_id,
            LikedFeeds.user_id == current_user.user_id
        ).first()
        
        if liked:
            # 좋아요 취소
            db.delete(liked)
            db.commit()
            is_liked = False
            message = "좋아요를 취소했습니다."
        else:
            # 좋아요 추가
            new_like = LikedFeeds(
                feed_id=feed_id,
                user_id=current_user.user_id,
                liked_at=datetime.now()
            )
            db.add(new_like)
            db.commit()
            is_liked = True
            message = "좋아요를 추가했습니다."
        
        # 좋아요 수 조회
        like_count = db.query(LikedFeeds).filter(LikedFeeds.feed_id == feed_id).count()
        
        return {
            "is_liked": is_liked,
            "like_count": like_count,
            "message": message
        }
    
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"좋아요 처리 중 오류가 발생했습니다: {str(e)}"
        )

# 피드 댓글 작성
@router.post("/{feed_id}/comments")
async def create_feed_comment(
    feed_id: int,
    content: str = Form(...),
    current_user: Users = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """피드 댓글 작성 API"""
    try:
        from app.models.feed_comments import FeedComments
        
        # 피드 존재 확인
        feed = db.query(Feeds).filter(Feeds.feed_id == feed_id).first()
        if not feed:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="피드를 찾을 수 없습니다."
            )
        
        # 댓글 작성
        new_comment = FeedComments(
            feed_id=feed_id,
            user_id=current_user.user_id,
            content=content,
            created_at=datetime.now()
        )
        db.add(new_comment)
        db.commit()
        db.refresh(new_comment)
        
        return {
            "comment_id": new_comment.comment_id,
            "feed_id": new_comment.feed_id,
            "user_id": new_comment.user_id,
            "content": new_comment.content,
            "created_at": new_comment.created_at,
            "user": {
                "user_id": current_user.user_id,
                "nickname": current_user.nickname,
                "profile_picture": current_user.profile_picture
            }
        }
    
    except HTTPException:
        raise
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"댓글 작성 중 오류가 발생했습니다: {str(e)}"
        )
